File: channel/src/workspace/tools/aes.go
package tools

import (
	"bytes"
	"crypto/cipher"
	"crypto/aes"
	"encoding/base64"
)


type Aes struct{
}


func (this *Aes) Encode(data ,key string)(string ,error){
	result, err := encrypt([]byte(data), []byte(key))
	if(err==nil){
		return base64.StdEncoding.EncodeToString(result),nil
	}
	return "",err
}

func (this *Aes) Decode(data ,key string)(string,error){
	mydata,err:=base64.StdEncoding.DecodeString(data);
	if(err==nil){
		origData, err := decrypt(mydata, []byte(key))
		if(err==nil){
			return string(origData),nil
		}else{
			return "",nil
		}
	}
	return "",err;
}

func main() {
	testAes()
}

func testAes() {
	//password:="sfe023f_9fd&fwfl"
	//data:="h"
	//encodedData,_:=AesEncode(data,password)
	//decodedData,_:=AesDecode(encodedData,password)
	//fmt.Println(encodedData)
	//fmt.Println(decodedData)

	/*// AES-128。key长度：16, 24, 32 bytes 对应 AES-128, AES-192, AES-256
	key := []byte("sfe023f_9fd&fwfl")
	result, err := encrypt([]byte("你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊"), key)
	if err != nil {
		panic(err)
	}

	base64Result:=base64.StdEncoding.EncodeToString(result)
	fmt.Println(base64Result)

	mydata,err:=base64.StdEncoding.DecodeString(base64Result);

	origData, err := decrypt(mydata, key)
	if err != nil {
		panic(err)
	}
	fmt.Println(string(origData))*/
}


func encrypt(origData, key []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	blockSize := block.BlockSize()
	origData = pKCS5Padding(origData, blockSize)
	// origData = zeroPadding(origData, block.BlockSize())
	blockMode := cipher.NewCBCEncrypter(block, key[:blockSize])
	crypted := make([]byte, len(origData))
	// 根据CryptBlocks方法的说明，如下方式初始化crypted也可以
	// crypted := origData
	blockMode.CryptBlocks(crypted, origData)
	return crypted, nil
}

func decrypt(crypted, key []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, err
	}
	blockSize := block.BlockSize()
	blockMode := cipher.NewCBCDecrypter(block, key[:blockSize])
	origData := make([]byte, len(crypted))
	// origData := crypted
	blockMode.CryptBlocks(origData, crypted)
	origData = pKCS5UnPadding(origData)
	// origData = zeroUnPadding(origData)
	return origData, nil
}

func zeroPadding(ciphertext []byte, blockSize int) []byte {
	padding := blockSize - len(ciphertext) % blockSize
	padtext := bytes.Repeat([]byte{0}, padding)
	return append(ciphertext, padtext...)
}

func zeroUnPadding(origData []byte) []byte {
	length := len(origData)
	unpadding := int(origData[length - 1])
	return origData[:(length-unpadding)]
}

func pKCS5Padding(ciphertext []byte, blockSize int) []byte {
	padding := blockSize - len(ciphertext) % blockSize
	padtext := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(ciphertext, padtext...)
}

func pKCS5UnPadding(origData []byte) []byte {
	length := len(origData)
	// 去掉最后一个字节 unpadding 次
	unpadding := int(origData[length - 1])
	return origData[:(length-unpadding)]
}

File: web/src/main/java/utils/AES.java
package utils;
import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import utils.base64.Base64;

public class AES {
    public static String decode(String data, String password) {
        AES aes = new AES(password);
        byte[] crypted = Base64.decodeToBytes(data);
        return new String(aes.decrypt(crypted));
    }
    /**
     * "sfe023f_9fd&fwfl"
     *
     * @param s
     * @param password
     * @return
     */
    public static String encode(String data, String password) {
        AES aes = new AES(password);
        byte[] crypted = aes.encrypt(data.getBytes());
        return Base64.encode(crypted);
    }
    /**
     * @param args
     * @throws Exception
     */
    public static void main(String[] args) throws Exception {
        AES aes = new AES("");
        String password = "sfe023f_9fd&fwfl";
        String data = "你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊你好啊";
        System.out.println(encode(data, password));
        System.out.println(decode(encode(data, password), password));
        /*  byte[] crypted = aes.encrypt(data.getBytes());
          System.out.println(Base64.encode(crypted));
          System.out.println(new String(aes.decrypt(crypted)));
          String encodedString = Base64.encode(data);
          System.out.println(encodedString);
          System.out.println(Base64.decode(encodedString));*/
    }
    private IvParameterSpec ivSpec;
    private SecretKeySpec keySpec;
    public AES(String key) {
        try {
            byte[] keyBytes = key.getBytes();
            byte[] buf = new byte[16];
            for (int i = 0; i < keyBytes.length && i < buf.length; i++) {
                buf[i] = keyBytes[i];
            }
            keySpec = new SecretKeySpec(buf, "AES");
            ivSpec = new IvParameterSpec(keyBytes);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    public byte[] decrypt(byte[] crypted) {
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.DECRYPT_MODE, keySpec, ivSpec);
            return cipher.doFinal(crypted);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
    public byte[] encrypt(byte[] origData) {
        try {
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, ivSpec);
            return cipher.doFinal(origData);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
}
